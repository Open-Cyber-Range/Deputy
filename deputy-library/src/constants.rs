use fancy_regex::Regex;
use lazy_static::lazy_static;
use parse_duration::parse;
use std::time::Duration;

pub const LOCKFILE: &str = "deputy.lock";
pub const LOCKFILE_TIMEOUT: &str = "5 minutes";
pub const LOCKFILE_SLEEP: &str = "250 milliseconds";

lazy_static! {
    pub static ref VALID_NAME: Regex = Regex::new(r#"^[a-zA-Z0-9_-]+$"#).unwrap();
    static ref LOCKFILE_TIMEOUT_DURATION: Duration =
        parse(LOCKFILE_TIMEOUT).expect("Error parsing lockfile timeout duration");
    pub static ref LOCKFILE_SLEEP_DURATION: Duration =
        parse(LOCKFILE_SLEEP).expect("Error parsing lockfile sleep duration");
    pub static ref LOCKFILE_TRIES: u64 =
        { LOCKFILE_TIMEOUT_DURATION.as_millis() / LOCKFILE_SLEEP_DURATION.as_millis() } as u64;
}

pub const SHA256_LENGTH: usize = 64;

pub const COMPRESSION_CHUNK_SIZE: usize = 131_072;

pub const PAYLOAD_CHUNK_SIZE: u64 = 8192;

pub const INDEX_REPOSITORY_BRANCH: &str = "master";
pub const INDEX_REPOSITORY_REMOTE: &str = "origin";
pub const CONFIGURATION_FOLDER_PATH_ENV_KEY: &str = "DEPUTY_CONFIG_FOLDER";
