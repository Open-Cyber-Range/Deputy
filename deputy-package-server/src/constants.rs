pub const fn default_page() -> u32 {
    1
}

pub const fn default_limit() -> u32 {
    20
}
