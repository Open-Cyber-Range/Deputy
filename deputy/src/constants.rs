use byte_unit::Byte;
use lazy_static::lazy_static;

lazy_static! {
    pub static ref SMALL_PACKAGE_LIMIT: u64 =
        u64::try_from(Byte::from_str("256kB").unwrap().get_bytes()).unwrap();
}

pub const CONFIGURATION_FILE_RELATIVE_PATH: &str = "configuration.toml";
pub const PACKAGE_TOML: &str = "package.toml";
pub const DEFAULT_REGISTRY_NAME: &str = "main-registry";

pub mod fetching {
    pub const DEFAULT_SAVE_PATH: &str = ".";
    pub const DEFAULT_PACKAGE_VERSION_REQUIREMENT: &str = "*";
}

pub mod endpoints {
    pub const SMALL_PACKAGE_UPLOAD_PATH: &str = "api/v1/package";
    pub const LARGE_PACKAGE_UPLOAD_PATH: &str = "api/v1/package/stream";
}
