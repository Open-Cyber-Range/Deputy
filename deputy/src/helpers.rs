use crate::constants::PACKAGE_TOML;
use anyhow::{anyhow, Error, Result};
use awc::error::PayloadError;
use bytes::Bytes;
use colored::Colorize;
use futures::{Stream, StreamExt};
use std::{io::Write, path::PathBuf};
use tempfile::TempDir;

pub fn find_toml(current_path: PathBuf) -> Result<PathBuf> {
    let mut toml_path = current_path.join(PACKAGE_TOML);
    if toml_path.is_file() {
        Ok(toml_path)
    } else if toml_path.pop() && toml_path.pop() {
        Ok(find_toml(toml_path)?)
    } else {
        Err(anyhow!("Could not find package.toml"))
    }
}

pub fn print_success_message(message: &str) {
    println!("{} {}", "Success:".green(), message);
}

pub fn print_error_message(error: Error) {
    eprintln!("{} {}", "Error:".red(), error);
}

pub fn create_temporary_package_download_path(
    package_name: &str,
    package_version: &str,
) -> Result<(String, TempDir)> {
    let temporary_directory = tempfile::Builder::new()
        .prefix(package_name)
        .rand_bytes(0)
        .tempdir()?;
    let file_name = temporary_directory.path().join(package_version);

    Ok((
        file_name
            .as_path()
            .to_str()
            .ok_or_else(|| anyhow!("Failed to create temporary path"))?
            .to_string(),
        temporary_directory,
    ))
}

pub async fn create_file_from_stream(
    stream: &mut (impl Stream<Item = Result<Bytes, PayloadError>> + Unpin + 'static),
    file_path: &str,
) -> Result<()> {
    let mut file = std::fs::File::create(file_path)?;
    while let Some(chunk) = stream.next().await {
        file.write_all(&chunk?)?;
        file.flush()?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use anyhow::Result;
    use tempfile::{Builder, TempDir};

    #[test]
    fn successfully_found_toml() -> Result<()> {
        let temp_dir = TempDir::new()?;
        let package_toml = Builder::new()
            .prefix("package")
            .suffix(".toml")
            .rand_bytes(0)
            .tempfile_in(&temp_dir)?;

        assert!(find_toml(temp_dir.path().to_path_buf())?.is_file());
        package_toml.close()?;
        temp_dir.close()?;
        Ok(())
    }

    #[test]
    fn creates_temporary_file_path() -> Result<()> {
        let package_name = "Shakespeare";
        let version = "0.5.0";
        let (temporary_path, temporary_directory) =
            create_temporary_package_download_path(package_name, version)?;
        insta::assert_display_snapshot!(temporary_path);
        temporary_directory.close()?;
        Ok(())
    }
}
