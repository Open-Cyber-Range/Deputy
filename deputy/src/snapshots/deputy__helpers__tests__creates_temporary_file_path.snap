---
source: deputy/src/helpers.rs
expression: temporary_path
---
/tmp/Shakespeare/0.5.0
